polynote-kernel/src/main/scala/polynote/kernel/util/DownloadableFile.scala

```diff
@@ -0,0 +1,61 @@
+package polynote.kernel.util
+
+import java.io.{File, FileInputStream, InputStream}
+import java.net.{HttpURLConnection, URI}
+import java.util.ServiceLoader
+
+import scala.collection.JavaConverters._
+import cats.effect.IO
+
+trait DownloadableFile {
+  def openStream: IO[InputStream]
+  def size: IO[Long]
+}
+
+trait DownloadableFileProvider {
+  def getFile(uri: URI): Option[DownloadableFile] = provide.lift(uri)
+
+  def provide: PartialFunction[URI, DownloadableFile]
```

Comment:
I get that the `PartialFunction` route looks more idiomatic scala-ish, and since we're dealing with a small number of these `DownloadableFileProvider`s it's totally fine... but in general I think the Java way is probably a more robust practice in general (i.e. the service interface returns a map of providers by protocol string, or something like that).

We're not going to blow the stack by `orElse`ing a couple of these `PartialFunction`s, but it still seems weird to go through every single one of them and say "How 'bout this URI?" instead of them registering their capability by the obvious thing that will determine it (the protocol)

I think it's still fine here, though. Just a general remark to be cautious about over-scalifying

Reply:
Yeah I did get a bit overexcited here with partial functions and lifts and unapplies :P 

I did consider having a `def protocols: Seq[String]` in `DownloadableFileProvider` which I think is more akin to what you're describing. 

Alternatively, do you think using something like `collectFirst` in `DownloadableFileProvider.getFile` would address that concern? It's a bit uglier but something like: 

```scala
  def getFile(uri: URI): Option[DownloadableFile] = providers.collectFirst {
    case p if p.getFile(uri).isDefined => p.getFile(uri).get
  }
```

Reply:
ok, after playing around with the idea I added `def protocols: Seq[String]` and `collectFirst` and I think it looks nicer now.